src/components/shared/SourceView.js

```diff
@@ -0,0 +1,170 @@
+      let maxLineNumber = sourceLines.length;
+      if (maxLineNumber <= 1) {
+        // We probably don't have the true source code yet, and don't really know
+        // the true number of lines in this file.
+        // Derive a maximum line number from the timings.
+        // Add a bit of space at the bottom (10 rows) so that the scroll position
+        // isn't too constrained - if the last known line is chosen as the "hot spot",
+        // this extra space allows us to display it in the top half of the viewport,
+        // if the viewport is small enough.
+        maxLineNumber = Math.max(1, ...timings.totalLineHits.keys()) + 10;
+      }
+      return range(1, maxLineNumber + 1);
+    },
+    { limit: 1 }
+  );
+
+  _computeMaxLineLengthMemoized = memoize(
+    (sourceLines: string[]): number =>
+      sourceLines.reduce(
+        (prevMaxLen, line) => Math.max(prevMaxLen, line.length),
+        0
+      ),
+    { limit: 1 }
+  );
+
+  _renderRow = (lineNumber: LineNumber, index: number, columnIndex: number) => {
+    const { rowHeight, timings } = this.props;
+    // React converts height into 'px' values, while lineHeight is valid in
+    // non-'px' units.
+    const rowHeightStyle = { height: rowHeight, lineHeight: `${rowHeight}px` };
+
+    const total = timings.totalLineHits.get(lineNumber);
+    const self = timings.selfLineHits.get(lineNumber);
+    const isNonZero = !!total || !!self;
+
+    if (columnIndex === 0) {
+      return (
+        <div
+          className={classNames('sourceViewRow', 'sourceViewRowFixedColumns', {
+            sourceViewRowNonZero: isNonZero,
+          })}
+          style={rowHeightStyle}
+        >
+          <span className="sourceViewRowColumn sourceViewFixedColumn total">
+            {total}
+          </span>
+          <span className="sourceViewRowColumn sourceViewFixedColumn self">
+            {self}
+          </span>
+          <span className="sourceViewRowColumn sourceViewFixedColumn lineNumber">
+            {lineNumber}
+          </span>
+        </div>
+      );
+    }
+
+    const sourceLines = this._getSourceLines();
+    const line = index < sourceLines.length ? sourceLines[index] : '';
+
+    return (
+      <div
+        className={classNames('sourceViewRow', 'sourceViewRowScrolledColumns', {
+          sourceViewRowNonZero: isNonZero,
+        })}
+        style={rowHeightStyle}
+        key={index}
+      >
+        <code>{line}</code>
+      </div>
+    );
+  };
+
+  _getSourceLines(): string[] {
+    return this._computeSourceLinesMemoized(this.props.source);
+  }
+
+  _getItems(): LineNumber[] {
+    const { timings } = this.props;
+    return this._computeAllLineNumbersMemoized(this._getSourceLines(), timings);
+  }
+
+  render() {
+    const { rowHeight, disableOverscan } = this.props;
+    const sourceLines = this._getSourceLines();
+    const maxLength = this._computeMaxLineLengthMemoized(sourceLines);
+    const CHAR_WIDTH_ESTIMATE = 8; // css pixels
```

Comment:
Hmm, good question. I guess it depends on the font. We should be able to make it exact, because we're using a fixed-width font so every letter has the same width.